### PR TITLE
Feature/question area

### PR DESCRIPTION
**Pull request by VadimZharikov**

### CODE REVIEW

File: AIS/AIS.API/Controllers/QuestionSetController.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+using AIS.BLL.Models;
+using AutoMapper;
+using Microsoft.AspNetCore.Mvc;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace AIS.API.Controllers
+{
+    [Route("api/[controller]")]
```
Comment (PaShkadun) [inline AIS/AIS.API/Controllers/QuestionSetController.cs:13]: We have const for this route.

File: AIS/AIS.API/Controllers/TrueAnswersController.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            var TrueAnswers = await _trueAnswerService.Get(ct);
+            return _mapper.Map<IEnumerable<TrueAnswerViewModel>>(TrueAnswers);
+        }
+
+        [HttpPost]
+        public async Task<TrueAnswerViewModel> AddTrueAnswer(TrueAnswerViewModel TrueAnswer, CancellationToken ct)
+        {
+            return _mapper.Map<TrueAnswerViewModel>(
+                await _trueAnswerService.Add(_mapper.Map<TrueAnswer>(TrueAnswer), ct)
+                );
```
Comment (PaShkadun) [inline AIS/AIS.API/Controllers/TrueAnswersController.cs:45]: Do it like in method in bottom.

**Review by PaShkadun — DISMISSED** [review]

**Review by PaShkadun — APPROVED** [review]

File: AIS/AIS.API/Controllers/QuestionsController.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        public QuestionsController(IGenericService<Question> questionService, IMapper mapper)
+        {
+            this._questionService = questionService;
+            this._mapper = mapper;
+        }
+
+        [HttpGet(EndpointConstants.IdTemplate)]
+        public async Task<QuestionViewModel> GetQuestion(int id, CancellationToken ct)
+        {
+            var Question = _mapper.Map<Question, QuestionViewModel>(await _questionService.GetById(id, ct));
```
[the remark this answers is not in the document]
Comment (Blackolf931) [reply AIS/AIS.API/Controllers/QuestionsController.cs:29]: you can make it easier "_mapper.Map<'QuestionViewModel'>"(param)

File: AIS/AIS.API/Controllers/QuestionAreaController.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        public QuestionAreaController(IGenericService<QuestionArea> questionAreaService, IMapper mapper)
+        {
+            _questionAreaService = questionAreaService;
+            _mapper = mapper;
+        }
+
+        [HttpGet(EndpointConstants.IdTemplate)]
+        public async Task<QuestionAreaViewModel> GetQuestionArea(int id, CancellationToken ct)
+        {
+            var questionArea = _mapper.Map<QuestionArea, QuestionAreaViewModel>(await _questionAreaService.GetById(id, ct));
```
Comment (PaShkadun) [inline AIS/AIS.API/Controllers/QuestionAreaController.cs:29]: You can do it easy (mapping).

File: AIS/AIS.API/Controllers/QuestionAreaController.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        [HttpPost]
+        public async Task<QuestionAreaViewModel> AddQuestionArea(QuestionAreaAddViewModel questionArea, CancellationToken ct)
+        {
+            return _mapper.Map<QuestionAreaViewModel>(
+                await _questionAreaService.Add(_mapper.Map<QuestionArea>(questionArea), ct)
+            );
+        }
+
+        [HttpPut]
+        public async Task<QuestionAreaViewModel> UpdateQuestionArea(QuestionAreaViewModel questionArea, CancellationToken ct)
```
Comment (PaShkadun) [inline AIS/AIS.API/Controllers/QuestionAreaController.cs:49]: You need use short model in parameters.

File: AIS/AIS.API/Controllers/QuestionSetController.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        public QuestionSetController(IGenericService<QuestionSet> questionSetService, IMapper mapper)
+        {
+            _questionSetService = questionSetService;
+            _mapper = mapper;
+        }
+
+        [HttpGet(EndpointConstants.IdTemplate)]
+        public async Task<QuestionSetViewModel> GetQuestionSet(int id, CancellationToken ct)
+        {
+            var questionSet = _mapper.Map<QuestionSet, QuestionSetViewModel>(await _questionSetService.GetById(id, ct));
```
Comment (PaShkadun) [inline AIS/AIS.API/Controllers/QuestionSetController.cs:29]: Do it easy.

File: AIS/AIS.API/Mapper/MappingProfile.cs [hunk 7]
```diff
@@ -17,12 +18,20 @@ public MappingProfile()
             CreateMap<ChangeIntervieweeViewModel, Interviewee>();
             CreateMap<Company, CompanyViewModel>();
             CreateMap<ChangeCompanyViewModel, Company>();
+            CreateMap<QuestionArea, QuestionAreaViewModel>().ReverseMap();
+            CreateMap<QuestionSet, QuestionSetViewModel>().ReverseMap();
+            CreateMap<Question, QuestionViewModel>().ReverseMap();
+            CreateMap<TrueAnswer, TrueAnswerViewModel>().ReverseMap();
```
Comment (Blackolf931) [inline AIS/AIS.API/Mapper/MappingProfile.cs:24]: Do you really need it? (ReverseMap())

File: AIS/AIS.API/Mapper/MappingProfile.cs [hunk 8]
```diff
@@ -17,12 +18,20 @@ public MappingProfile()
...
             CreateMap<Session, SessionViewModel>();
             CreateMap<SessionAddViewModel, Session>();
             CreateMap<SessionUpdateViewModel, Session>();
             CreateMap<Company, ShortCompanyViewModel>();
             CreateMap<Employee, ShortEmployeeViewModel>();
             CreateMap<Interviewee, ShortIntervieweeViewModel>();
+            CreateMap<QuestionArea, QuestionAreaAddViewModel>().ReverseMap();
+            CreateMap<QuestionSet, QuestionSetAddViewModel>().ReverseMap();
+            CreateMap<Question, QuestionAddViewModel>().ReverseMap();
+            CreateMap<TrueAnswer, TrueAnswerAddViewModel>().ReverseMap();
```
Comment (Blackolf931) [inline AIS/AIS.API/Mapper/MappingProfile.cs:34]: The same

File: AIS/AIS.API/Controllers/QuestionSetController.cs [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        [HttpPost]
+        public async Task<QuestionSetViewModel> AddQuestionSet(QuestionSetAddViewModel questionSet, CancellationToken ct)
+        {
+            return _mapper.Map<QuestionSetViewModel>(
+                await _questionSetService.Add(_mapper.Map<QuestionSet>(questionSet), ct)
+            );
+        }
+
+        [HttpPut]
+        public async Task<QuestionSetViewModel> UpdateQuestionSet(QuestionSetAddViewModel questionSet, CancellationToken ct)
```
Comment (PaShkadun) [inline AIS/AIS.API/Controllers/QuestionSetController.cs:49]: Is it addViewModel in update method?

File: AIS/AIS.API/Controllers/QuestionsController.cs (hunk 3 above)
Comment (PaShkadun) [inline AIS/AIS.API/Controllers/QuestionsController.cs:29]: Do it easy.

File: AIS/AIS.API/Controllers/QuestionsController.cs [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        [HttpPost]
+        public async Task<QuestionViewModel> AddQuestion(QuestionAddViewModel Question, CancellationToken ct)
+        {
+            return _mapper.Map<QuestionViewModel>(
+                await _questionService.Add(_mapper.Map<Question>(Question), ct)
+            );
+        }
+
+        [HttpPut]
+        public async Task<QuestionViewModel> UpdateQuestion(QuestionViewModel Question, CancellationToken ct)
```
Comment (PaShkadun) [inline AIS/AIS.API/Controllers/QuestionsController.cs:49]: Short model.

**Review by Blackolf931 — CHANGES_REQUESTED** [review]

File: AIS/AIS.API/Controllers/TrueAnswersController.cs [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        public TrueAnswersController(IGenericService<TrueAnswer> trueAnswerService, IMapper mapper)
+        {
+            this._trueAnswerService = trueAnswerService;
+            this._mapper = mapper;
+        }
+
+        [HttpGet(EndpointConstants.IdTemplate)]
+        public async Task<TrueAnswerViewModel> GetTrueAnswer(int id, CancellationToken ct)
+        {
+            var trueAnswer = _mapper.Map<TrueAnswer, TrueAnswerViewModel>(await _trueAnswerService.GetById(id, ct));
```
Comment (PaShkadun) [inline AIS/AIS.API/Controllers/TrueAnswersController.cs:29]: Do it easy.

File: AIS/AIS.API/Controllers/TrueAnswersController.cs [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        [HttpPost]
+        public async Task<TrueAnswerViewModel> AddTrueAnswer(TrueAnswerAddViewModel trueAnswer, CancellationToken ct)
+        {
+            return _mapper.Map<TrueAnswerViewModel>(
+                await _trueAnswerService.Add(_mapper.Map<TrueAnswer>(trueAnswer), ct)
+            );
+        }
+
+        [HttpPut]
+        public async Task<TrueAnswerViewModel> UpdateTrueAnswer(TrueAnswerViewModel trueAnswer, CancellationToken ct)
```
Comment (PaShkadun) [inline AIS/AIS.API/Controllers/TrueAnswersController.cs:49]: Short model.

**Review by PaShkadun — CHANGES_REQUESTED** [review]

File: AIS/AIS.API.Tests/Controllers/EmployeeControllerTests.cs [hunk 13]
```diff
@@ -41,19 +39,19 @@ public async Task Add_WhenControllerHasData_ShouldReturnValidModel()
...
+            var controller = new EmployeeController(_mockMapper.Object, _mockService.Object, _mockValidator.Object);
 
             // Act
             var result = await controller.Add(inputEmployeeViewModel, default);
 
             // Assert
-            inputEmployeeViewModel.Name.ShouldBeEquivalentTo(result.Name);
-            inputEmployeeViewModel.CompanyId.ShouldBeEquivalentTo(result.CompanyId);
+            Assert.Equal(inputEmployeeViewModel.Name, result.Name);
+            Assert.Equal(inputEmployeeViewModel.CompanyId, result.CompanyId);
```
Comment (PaShkadun) [inline AIS/AIS.API.Tests/Controllers/EmployeeControllerTests.cs:54]: Why did you delete shouldly and did it with Assert?

File: AIS/AIS.API.Tests/Controllers/EmployeeControllerTests.cs [hunk 14]
```diff
@@ -163,16 +148,16 @@ public void Delete_WhenContollerHasData_NoReturn()
...
             // Act
-            Action act = () => controller.Delete(expected.Id, default);
+            Func<Task> act = async () => await controller.Delete(expected.Id, default);
 
             // Assert
-            act.Should().NotThrow();
+            act.Should().NotThrowAsync();
         }
     }
 }
```
Comment (Blackolf931) [inline AIS/AIS.API.Tests/Controllers/EmployeeControllerTests.cs:163]: Why you change this file?

File: AIS/AIS.DAL/Repositories/IntervieweeRepository.cs [hunk 15]
```diff
@@ -29,8 +30,23 @@ public override async Task<IEnumerable<IntervieweeEntity>> Get(Expression<Func<I
         public override async Task<IntervieweeEntity> GetById(int id, CancellationToken ct)
         {
             var entity = await _dbSet.Include(x => x.Company).FirstOrDefaultAsync(x => x.Id == id, ct);
-
             return entity;
         }
+
+        protected override IQueryable<IntervieweeEntity> Query(bool eager = false)
```
Comment (PaShkadun) [inline AIS/AIS.DAL/Repositories/IntervieweeRepository.cs:36]: Why? Is it necessary in this repository? It doesn't use.

File: AIS/AIS.API.Tests/Controllers/EmployeeControllerTests.cs [hunk 16]
```diff
@@ -76,19 +74,19 @@ public async Task Update_WhenControllerHasData_ShouldReturnValidModel()
...
+            var controller = new EmployeeController(_mockMapper.Object, _mockService.Object, _mockValidator.Object);
 
             // Act
             var result = await controller.Update(inputEmployeeViewModel, 2, default);
 
             // Assert
-            inputEmployeeViewModel.CompanyId.ShouldBeEquivalentTo(result.CompanyId);
-            inputEmployeeViewModel.Name.ShouldBeEquivalentTo(result.Name);
+            Assert.Equal(inputEmployeeViewModel.CompanyId, result.CompanyId);
+            Assert.Equal(inputEmployeeViewModel.Name, result.Name);
```
Comment (PaShkadun) [inline AIS/AIS.API.Tests/Controllers/EmployeeControllerTests.cs:89]: The same.

**Review by PaShkadun — CHANGES_REQUESTED** [review]

File: AIS/AIS.API/Controllers/QuestionsController.cs [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+        public async Task<QuestionViewModel> GetQuestion(int id, CancellationToken ct)
+        {
+            var Question = _mapper.Map<Question, QuestionViewModel>(await _questionService.GetById(id, ct));
+            return Question;
+        }
+
+        [HttpGet]
+        public async Task<IEnumerable<QuestionViewModel>> GetQuestions(CancellationToken ct)
+        {
+            var Questions = await _questionService.Get(ct);
```
[the remark this answers is not in the document]
Comment (Blackolf931) [reply AIS/AIS.API/Controllers/QuestionsController.cs:36]: image

File: AIS/AIS.API/Controllers/QuestionsController.cs [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+            return _mapper.Map<IEnumerable<QuestionViewModel>>(Questions);
+        }
+
+        [HttpPost]
+        public async Task<QuestionViewModel> AddQuestion(QuestionAddViewModel Question, CancellationToken ct)
```
Comment (Blackolf931) [inline AIS/AIS.API/Controllers/QuestionsController.cs:41]: Question => question

File: AIS/AIS.API/Controllers/QuestionsController.cs [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            return _mapper.Map<QuestionViewModel>(
+                await _questionService.Add(_mapper.Map<Question>(Question), ct)
+            );
+        }
+
+        [HttpPut]
+        public async Task<QuestionViewModel> UpdateQuestion(QuestionUpdateViewModel Question, CancellationToken ct)
```
Comment (Blackolf931) [inline AIS/AIS.API/Controllers/QuestionsController.cs:49]: The same

File: AIS/AIS.API/ViewModels/QuestionArea/QuestionAreaUpdateViewModel.cs [hunk 20]
```diff
@@ -0,0 +1,8 @@
+﻿namespace AIS.API.ViewModels.QuestionArea
+{
+    public class QuestionAreaUpdateViewModel
+    {
+        public int Id { get; set; }
```
Comment (Blackolf931) [inline AIS/AIS.API/ViewModels/QuestionArea/QuestionAreaUpdateViewModel.cs:5]: We don't use Id prop this

File: AIS/AIS.API/ViewModels/QuestionSet/QuestionSetUpdateViewModel.cs [hunk 21]
```diff
@@ -0,0 +1,9 @@
+﻿namespace AIS.API.ViewModels.QuestionSet
+{
+    public class QuestionSetUpdateViewModel
+    {
+        public int Id { get; set; }
```
Comment (Blackolf931) [inline AIS/AIS.API/ViewModels/QuestionSet/QuestionSetUpdateViewModel.cs:5]: The same

File: AIS/AIS.API/ViewModels/TrueAnswer/TrueAnswerUpdateViewModel.cs [hunk 22]
```diff
@@ -0,0 +1,9 @@
+﻿namespace AIS.API.ViewModels.TrueAnswer
+{
+    public class TrueAnswerUpdateViewModel
+    {
+        public int Id { get; set; }
```
Comment (Blackolf931) [inline AIS/AIS.API/ViewModels/TrueAnswer/TrueAnswerUpdateViewModel.cs:5]: The same

File: AIS/AIS.API.Tests/Controllers/EmployeeControllerTests.cs (hunk 14 above)
Comment (Blackolf931) [reply AIS/AIS.API.Tests/Controllers/EmployeeControllerTests.cs:163]: Maybe you copy this file from main

**Review by Blackolf931 — CHANGES_REQUESTED** [review]

File: AIS/AIS.DAL/Repositories/IntervieweeRepository.cs (hunk 15 above)
Comment (VadimZharikov) [reply AIS/AIS.DAL/Repositories/IntervieweeRepository.cs:36]: Check the base class

File: AIS/AIS.API.Tests/Controllers/EmployeeControllerTests.cs (hunk 14 above)
Comment (VadimZharikov) [reply AIS/AIS.API.Tests/Controllers/EmployeeControllerTests.cs:163]: methods support async implementation

**Review by nantonov — CHANGES_REQUESTED** [review]

File: AIS/AIS.API/Mapper/MappingProfile.cs [hunk 23]
```diff
@@ -21,24 +21,24 @@ public MappingProfile()
...
             CreateMap<Company, CompanyViewModel>();
             CreateMap<ChangeCompanyViewModel, Company>();
-            CreateMap<QuestionArea, QuestionAreaViewModel>();
-            CreateMap<QuestionSet, QuestionSetViewModel>();
-            CreateMap<Question, QuestionViewModel>();
-            CreateMap<TrueAnswer, TrueAnswerViewModel>();
+            CreateMap<QuestionArea, QuestionAreaViewModel>().ReverseMap();
+            CreateMap<QuestionSet, QuestionSetViewModel>().ReverseMap();
+            CreateMap<Question, QuestionViewModel>().ReverseMap();
+            CreateMap<TrueAnswer, TrueAnswerViewModel>().ReverseMap();
```
Comment (PaShkadun) [inline AIS/AIS.API/Mapper/MappingProfile.cs]: Why?

File: AIS/AIS.API/Mapper/MappingProfile.cs [hunk 24]
```diff
@@ -21,24 +21,24 @@ public MappingProfile()
...
-            CreateMap<QuestionUpdateViewModel, Question>();
-            CreateMap<TrueAnswerUpdateViewModel, TrueAnswer>();
+            CreateMap<QuestionAreaAddViewModel, QuestionArea>().ReverseMap();
+            CreateMap<QuestionSetAddViewModel, QuestionSet>().ReverseMap();
+            CreateMap<QuestionAddViewModel, Question>().ReverseMap();
+            CreateMap<TrueAnswerAddViewModel, TrueAnswer>().ReverseMap();
+            CreateMap<QuestionAreaUpdateViewModel, QuestionArea>().ReverseMap();
+            CreateMap<QuestionSetUpdateViewModel, QuestionSet>().ReverseMap();
+            CreateMap<QuestionUpdateViewModel, Question>().ReverseMap();
+            CreateMap<TrueAnswerUpdateViewModel, TrueAnswer>().ReverseMap();
```
Comment (PaShkadun) [inline AIS/AIS.API/Mapper/MappingProfile.cs]: Why?

File: AIS/AIS.API/Mapper/MappingProfile.cs (hunk 23 above)
Comment (VadimZharikov) [reply AIS/AIS.API/Mapper/MappingProfile.cs]: to map navigational properties dependencies

File: AIS/AIS.API/Mapper/MappingProfile.cs (hunk 24 above)
Comment (VadimZharikov) [reply AIS/AIS.API/Mapper/MappingProfile.cs]: to map navigational properties dependencies

**Review by PaShkadun — APPROVED** [review]

**Review by Blackolf931 — APPROVED** [review]

**Review by nantonov — APPROVED** [review]

**Pull request merged**